src/cppgc_helpers.h

```diff
@@ -0,0 +1,108 @@
+#ifndef SRC_CPPGC_HELPERS_H_
+#define SRC_CPPGC_HELPERS_H_
+
+#if defined(NODE_WANT_INTERNALS) && NODE_WANT_INTERNALS
+
+#include <type_traits>  // std::remove_reference
+#include "cppgc/garbage-collected.h"
+#include "cppgc/name-provider.h"
+#include "env.h"
+#include "memory_tracker.h"
+#include "v8-cppgc.h"
+#include "v8.h"
+
+namespace node {
+
+/**
+ * This is a helper mixin with a BaseObject-like interface to help
+ * implementing wrapper objects managed by V8's cppgc (Oilpan) library.
+ * cppgc-manged objects in Node.js internals should extend this mixin,
+ * while non-cppgc-managed objects typically extend BaseObject - the
+ * latter are being migrated to be cppgc-managed wherever it's beneficial
+ * and practical. Typically cppgc-managed objects are more efficient to
+ * keep track of (which lowers initialization cost) and work better
+ * with V8's GC scheduling.
+ *
+ * A cppgc-managed native wrapper should look something like this, note
+ * that per cppgc rules, cppgc::GarbageCollected<> must be the left-most
+ * base class.
+ *
+ * class Klass final : public cppgc::GarbageCollected<Klass>,
+ *                     public cppgc::NameProvider,
+ *                     public CppgcMixin {
+ *  public:
+ *   SET_CPPGC_NAME(Klass)  // Sets the heap snapshot name to "Node / Klass"
+ *   void Trace(cppgc::Visitor* visitor) const final {
+ *     CppgcMixin::Trace(visitor);
+ *     visitor->Trace(...);  // Trace any additional owned traceable data
+ *   }
+ * }
+ */
+class CppgcMixin : public cppgc::GarbageCollectedMixin {
+ public:
+  // To help various callbacks access wrapper objects with different memory
+  // management, cppgc-managed objects share the same layout as BaseObjects.
+  enum InternalFields { kEmbedderType = 0, kSlot, kInternalFieldCount };
+
+  // The initialization cannot be done in the mixin constructor but has to be
+  // invoked from the child class constructor, per cppgc::GarbageCollectedMixin
+  // rules.
+  template <typename T>
+  void InitializeCppgc(T* ptr, Environment* env, v8::Local<v8::Object> obj) {
```

Comment:
I think the name `InitializeCppgc` suggests initializing the whole cppgc setup. This instance method just initialize the object and the instance. A name like `InitializeGarbageCollected` might be better?

```suggestion
  void InitializeGarbageCollected(T* ptr, Environment* env, v8::Local<v8::Object> obj) {
```

Reply:
Umm, I feel that `GarbageCollected` may still not be very clear about cppgc (could be interpreted as "initialize a JS reference that's garbage collected). Perhaps `CppgcWrapper` is easier to understand since that's what it is about.

Reply:
Actually this can just be called `Wrap()`, and invoked as `CppgcMixin::Wrap` to be clear.

Reply:
Yeah, sounds better

Reply:
I noticed that there is also a global `Unwrap` which is meant for `BaseObject`, so I moved that into `BaseObject` to disambiguate, so now `BaseObject::Unwrap` corresponds nicely to `CppgcMixin::Unwrap`.